Sources/StreamVideo/CallStateMachine/Stages/Call+JoiningStage.swift

```diff
@@ -0,0 +1,160 @@
+
+    /// Creates a joining stage for a call with the given input.
+    ///
+    /// - Parameters:
+    ///   - call: The call to be joined
+    ///   - input: The input containing join parameters
+    /// - Returns: A new `JoiningStage` instance
+    static func joining(
+        _ call: Call,
+        input: Context.JoinInput
+    ) -> Call.StateMachine.Stage {
+        JoiningStage(
+            .init(
+                call: call,
+                joinInput: input
+            )
+        )
+    }
+}
+
+extension Call.StateMachine.Stage {
+
+    /// Represents the joining stage in the call state machine.
+    final class JoiningStage: Call.StateMachine.Stage, @unchecked Sendable {
+        @Injected(\.streamVideo) private var streamVideo
+
+        private let disposableBag = DisposableBag()
+        private var iterations = 0
+
+        /// Creates a new joining stage with the provided context.
+        ///
+        /// - Parameter context: The call context containing necessary state
+        init(
+            _ context: Context
+        ) {
+            super.init(id: .joining, context: context)
+        }
+
+        /// Handles state transitions to the joining stage.
+        ///
+        /// Valid transitions:
+        /// - From: IdleStage, AcceptedStage, JoiningStage
+        /// - To: JoinedStage, ErrorStage
+        ///
+        /// - Parameter previousStage: The stage transitioning from
+        /// - Returns: Self if transition is valid, nil otherwise
+        override func transition(
+            from previousStage: Call.StateMachine.Stage
+        ) -> Self? {
+            switch previousStage.id {
+            case .idle, .accepted, .joining:
+                execute()
+                return self
+            default:
+                return nil
+            }
+        }
+
+        // MARK: - Private Helpers
+
+        /// Executes the call joining process asynchronously.
+        private func execute() {
+            Task { [weak self] in
+                guard let self else { return }
+
+                try? Task.checkCancellation()
+
+                guard
+                    let call = context.call,
+                    let input = context.joinInput
+                else {
+                    transitionErrorOrLog(ClientError("Invalid input to join call."))
+                    return
+                }
+
+                log.debug("Will attempt joining call id:\(call.callId) type:\(call.callType) with input: \(input).")
+
+                do {
+                    let response = try await executeJoin(call: call, input: input)
+                    transitionOrError(.joined(context, response: response))
+                } catch {
+                    var input = input
+                    input.currentNumberOfRetries += 1
+
+                    if input.currentNumberOfRetries < input.maxNumberOfRetries {
+                        let delay = UInt64(TimeInterval.random(in: input.retryDelayRange) * 1_000_000_000)
```

Comment:
why don't we use the retry policies we already have?

Reply:
The retry policies that accept a closure and obscure all internal transitions. I wanted to move the workload to the statemachine and stop using closures. So now our logs are pointing directly how the operation goes and it's easier to track and debug. It's worth mentioning that the retry and delay logic we have here is the same as on the `RetryPolicy.fastAndSimple`.

Reply:
Yes, but the logic itself is duplicated. The delay logic seems to be a static function in RetryPolicy, maybe we can just use that?

Reply:
Changed it to reuse the RetryPolicy struct (but not the executeTask that contains)